BlockVolume add extra arguments to create and reduce number of cluster locks

poornimag opened the pull request:
With this the block volumes can be created on thin arbiter, shard volumes. 

Signed-off-by: Poornima G <pgurusid@redhat.com>

poornimag commented:
@oshankkumar @aravindavk Please review the patch. Thanks

aravindavk commented:
add to whitelist

oshankkumar commented on plugins/blockvolume/hostvol/utils.go:
```diff
@@ -15,34 +15,70 @@ import (
 	log "github.com/sirupsen/logrus"
 )
 
+func filterSize(volinfo *volume.Volinfo, size uint64) bool {
```
size filter is already present, please use the existing one.

poornimag replied:
Done

oshankkumar commented on plugins/blockvolume/hostvol/utils.go:
```diff
@@ -15,34 +15,70 @@ import (
 // BlockSizeFilter returns a volume Filter, which will filter out volumes
 // haing block-hosting-available-size greater than give size.
-func BlockSizeFilter(size uint64) volume.Filter {
+func HostOptnsFilter(size uint64, h *HostingVolumeOptions) volume.Filter {
```
write applyArbiter and applyShard separately as a filter function and this function can be omitted.

oshankkumar replied:
you can add  FilterArbiter and FilterShard func in this file `github.com/gluster/glusterd2/glusterd2/volume/filters.go`

poornimag replied:
Done

oshankkumar commented on plugins/blockvolume/hostvol/utils.go:
```diff
@@ -15,34 +15,70 @@ import (
+func GetExistingBlockHostingVolume(size uint64, h *HostingVolumeOptions) (*volume.Volinfo, error) {
 	var (
-		filters     = []volume.Filter{volume.FilterBlockHostedVolumes, BlockSizeFilter(size)}
+		filters     = []volume.Filter{volume.FilterBlockHostedVolumes, HostOptnsFilter(size, h)}
```
use can use filters like []volume.Filter{volume.FilterBlockHostedVolumes, BlockSizeFilter(size),FilterArbiter,FilterShard}.
FilterArbiter,FilterShard need to be implemented

poornimag replied:
I thought of the same, but if i were to apply filter separately then each filter has to loop though the list of all volumes right? So wrote it like this, for each volume apply all filters required. If the number of volumes are huge, this looping of volumes for each filter adds up redundancy? But yeah block hosting volumes are generally less in number, hence not much difference may be. Thought?

oshankkumar replied:
right, if the number of volumes will be more then it will be a problem, okay then we can go with the current implementation.

poornimag replied:
Done

Madhu-1 commented on plugins/blockvolume/api/types.go:
```diff
@@ -11,9 +11,23 @@ type BlockVolumeInfo struct {
+	// HostVolShardSize represents the shard size of the block hosting volume
+	HostVolShardSize uint64 `json:"hostvolshardsize,omitempty"`
+	// Size represents Block Hosting Volume size in bytes
+	HostVolSize uint64 `json:"hostsize,omitempty"`
```
`%s/hostsize/hostvolsize`

Madhu-1 commented on plugins/blockvolume/api/types.go:
```diff
@@ -11,9 +11,23 @@ type BlockVolumeInfo struct {
 type BlockVolumeCreateRequest struct {
 	*BlockVolumeInfo
+	HostVolumeInfo
+	BlkType  string   `json:"blktype,omitempty"`
```
what is `blktype`?

poornimag replied:
It will contain the filesystem type the block will be formatted with, if its "none" then the file will not be formatted with any filesystem. The default is "xfs".

Madhu-1 replied:
can we make it as `block-type` for more readability and set a default value on server side based on the empty string will be a good close instead of "none"

poornimag replied:
Sure, may be will use "raw" instead of "none"?

Madhu-1 commented on plugins/blockvolume/blockprovider/gluster-virtblock/gluster_virtblock.go:
```diff
@@ -99,21 +99,33 @@ func (g *GlusterVirtBlk) CreateBlockVolume(name string, size uint64, hostVolume
-	if err != nil {
-		logger.WithError(err).Errorf("failed to format block file " + blockFileName)
-		return nil, err
+	if blockVolOpts.BlkType != "none" {
```
why it is compared with none? can this be compared with an empty string which default none for string?

poornimag replied:
"none" is used to represent that no formatting is required, if its "" then the default is "xfs".

Madhu-1 replied:
sorry, am not aware of no formatting us required use case can you explain?

poornimag replied:
https://kubernetes.io/docs/concepts/storage/persistent-volumes/#volume-mode  This one? VolumeMode can be raw as well, not sure about further deatils.

Madhu-1 commented on plugins/blockvolume/blockprovider/gluster-virtblock/gluster_virtblock.go:
```diff
@@ -99,21 +99,33 @@ func (g *GlusterVirtBlk) CreateBlockVolume(name string, size uint64, hostVolume
+		fsType := blockVolOpts.BlkType
+		err = utils.ExecuteCommandRun(fmt.Sprintf("mkfs.%s", fsType), "-f", blockFileName) //nolint: gosec
+		if err != nil {
+			logger.WithError(err).Errorf("failed to format block file " + blockFileName + " with filesystem " + fsType)
```
formatting required for `Errorf` 
```
logger.WithError(err).Errorf("failed to format block file %s with filesystem %s" , blockFileName, fsType)
```

poornimag replied:
Done

Madhu-1 commented on plugins/blockvolume/blockprovider/gluster-virtblock/gluster_virtblock.go:
```diff
@@ -99,21 +99,33 @@ func (g *GlusterVirtBlk) CreateBlockVolume(name string, size uint64, hostVolume
-	if err = clusterLocks.Lock(hostVolume); err != nil {
-		logger.WithError(err).Error("error in acquiring cluster lock")
-		return nil, err
+	for {
```
infinite loop could cause an issue, which may lead to http timeout

poornimag replied:
So we try for few iterations, and then return with LockTimeout? Then the csi should handle this error and retry?

Madhu-1 replied:
if volume creation fails, CSI will do a retry

Madhu-1 commented on plugins/blockvolume/blockprovider/gluster-virtblock/gluster_virtblock.go:
```diff
@@ -142,9 +154,18 @@ func delBlockEntry(hostName string, name string) error {
-	if err := clusterLocks.Lock(hostName); err != nil {
-		log.WithError(err).Error("error in acquiring cluster lock")
-		return err
+	for {
```
same here

poornimag replied:
Done

Madhu-1 commented on plugins/blockvolume/hostvol/hostvol_manager.go:
```diff
@@ -55,21 +55,35 @@ func (g *glusterVolManager) GetHostingVolumesInUse() []*volume.Volinfo {
-	if err := clusterLocks.Lock(path.Join(globalLockID, name)); err != nil {
-		return nil, err
+	for {
+		err := clusterLocks.Lock(path.Join(globalLockID))
```
same here

poornimag commented:
> what if the hostvolume is different? then we should not block.

Even when the host volume is different we still need global lock for creating host volume. Consider a situation, where there are two block create requests issued simultaneously, then we have two block host host volumes created. Hence we need a global lock for creating block host volume. The only time we may not need it, is when the bhv is passed a input in the block create request. Is that the case you are talking about?

oshankkumar commented on plugins/blockvolume/hostvol/hostvol_manager.go:
```diff
@@ -55,21 +55,35 @@ func (g *glusterVolManager) GetHostingVolumesInUse() []*volume.Volinfo {
-	if err := clusterLocks.Lock(path.Join(globalLockID, name)); err != nil {
-		return nil, err
+	for {
+		err := clusterLocks.Lock(path.Join(globalLockID))
```
if host volume name is empty, then `path.Join(globalLockID, name)` resolves in `globalLockID` only. So I don't think this change is required.

poornimag replied:
Ah ok. @aravindavk The block host volume creation is already serialized, then how can there be stale bricks? Before creating bhv it checks for alerady existing bhv under lock. So its not root cause for stale bricks?

oshankkumar replied:
parallel req for BHV creation is already taken care of. This is not the root cause for any stale bricks.

poornimag commented:
@Madhu-1 @oshankkumar Have addressed the comments, and got rid of patches with infinite loop for lock and bhv lock id change. Any other changes required?

Madhu-1 commented on plugins/blockvolume/blockprovider/gluster-virtblock/gluster_virtblock.go:
```diff
@@ -179,9 +155,10 @@ func (g *GlusterVirtBlk) DeleteBlockVolume(name string, options ...blockprovider
+	hostDir, err := mountHost(g, hostName)
 	if err != nil {
-		log.WithError(err).Errorf("error mounting block hosting volume :" + blkVol.HostVolume())
+		log.WithError(err).Errorf("error mounting block hosting volume :" + hostName)
```
formatting  is required for Errorf

oshankkumar commented:
LGTM

Madhu-1 commented on glusterd2/volume/volume-utils.go:
```diff
@@ -399,3 +400,28 @@ func CleanBricksLoop(volinfo *Volinfo) error {
+	req.Options["features/shard"] = "on"
+	req.Options["features/shard.shard-block-size"] = strconv.FormatUint(shardSize, 10)
+	req.AllowAdvanced = true
+	return nil
```
what's the use of returning an error here? it's always nil

poornimag replied:
Nothing, just to keep it similar to AddThinArbiter.., Will change

poornimag commented:
@Madhu-1 Anyother changes? Thanks

poornimag commented:
@aravindavk Can you merge this patch?
